File: 1000-1099/1081. Smallest Subsequence of Distinct Characters/Solution.java
//Approach-1 (Greedy + Monotonic Stack)
//T.C : O(n)
//S.C : O(1)

class Solution {

  public String smallestSubsequence(
      String s) {

    int[] lastIndex = new int[26];

    boolean[] inStack = new boolean[26];

    StringBuilder stack = new StringBuilder();

    for (int i = 0; i < s.length(); i++) {

      lastIndex[s.charAt(i) - 'a'] = i;
    }

    for (int i = 0; i < s.length(); i++) {

      int current = s.charAt(i) - 'a';

      if (inStack[current]) {

        continue;
      }

      while (stack.length() > 0) {

        int top = stack.charAt(
            stack.length() - 1) - 'a';

        if (top > current &&
            lastIndex[top] > i) {

          stack.deleteCharAt(
              stack.length() - 1);

          inStack[top] = false;

        } else {

          break;
        }
      }

      stack.append(
          s.charAt(i));

      inStack[current] = true;
    }

    return stack.toString();
  }
}

/*
 * //Approach-2 (Using Stack)
 * //T.C : O(n)
 * //S.C : O(n)
 * 
 * import java.util.Stack;
 * 
 * class Solution {
 * 
 * public String smallestSubsequence(
 * String s
 * ) {
 * 
 * int[] lastIndex =
 * new int[26];
 * 
 * boolean[] visited =
 * new boolean[26];
 * 
 * Stack<Character> stack =
 * new Stack<>();
 * 
 * for (
 * int i = 0;
 * i < s.length();
 * i++
 * ) {
 * 
 * lastIndex[
 * s.charAt(i) - 'a'
 * ] = i;
 * }
 * 
 * for (
 * int i = 0;
 * i < s.length();
 * i++
 * ) {
 * 
 * char current =
 * s.charAt(i);
 * 
 * if (
 * visited[current - 'a']
 * ) {
 * 
 * continue;
 * }
 * 
 * while (
 * !stack.isEmpty() &&
 * stack.peek() > current &&
 * lastIndex[
 * stack.peek() - 'a'
 * ] > i
 * ) {
 * 
 * visited[
 * stack.pop() - 'a'
 * ] = false;
 * }
 * 
 * stack.push(
 * current
 * );
 * 
 * visited[
 * current - 'a'
 * ] = true;
 * }
 * 
 * StringBuilder answer =
 * new StringBuilder();
 * 
 * while (
 * !stack.isEmpty()
 * ) {
 * 
 * answer.append(
 * stack.pop()
 * );
 * }
 * 
 * return answer
 * .reverse()
 * .toString();
 * }
 * }
 */

File: 1000-1099/1081. Smallest Subsequence of Distinct Characters/Solution.cpp
//Approach-1 (Greedy + Monotonic Stack)
//T.C : O(n)
//S.C : O(1)

class Solution {
public:

    string smallestSubsequence(
        string s
    ) {

        vector<int> lastIndex(
            26
        );

        vector<bool> inStack(
            26,
            false
        );

        string stack = "";

        for (
            int i = 0;
            i < s.size();
            i++
        ) {

            lastIndex[
                s[i] - 'a'
            ] = i;
        }

        for (
            int i = 0;
            i < s.size();
            i++
        ) {

            int current =
                s[i] - 'a';

            if (
                inStack[current]
            ) {

                continue;
            }

            while (
                !stack.empty()
            ) {

                int top =
                    stack.back() - 'a';

                if (
                    top > current &&
                    lastIndex[top] > i
                ) {

                    inStack[top] =
                        false;

                    stack.pop_back();

                } else {

                    break;
                }
            }

            stack.push_back(
                s[i]
            );

            inStack[current] =
                true;
        }

        return stack;
    }
};


/*
//Approach-2 (Using std::stack)
//T.C : O(n)
//S.C : O(n)

class Solution {
public:

    string smallestSubsequence(
        string s
    ) {

        vector<int> lastIndex(
            26
        );

        vector<bool> visited(
            26,
            false
        );

        stack<char> st;

        for (
            int i = 0;
            i < s.size();
            i++
        ) {

            lastIndex[
                s[i] - 'a'
            ] = i;
        }

        for (
            int i = 0;
            i < s.size();
            i++
        ) {

            char current =
                s[i];

            if (
                visited[current - 'a']
            ) {

                continue;
            }

            while (
                !st.empty() &&
                st.top() > current &&
                lastIndex[st.top() - 'a'] > i
            ) {

                visited[
                    st.top() - 'a'
                ] = false;

                st.pop();
            }

            st.push(
                current
            );

            visited[
                current - 'a'
            ] = true;
        }

        string answer = "";

        while (
            !st.empty()
        ) {

            answer +=
                st.top();

            st.pop();
        }

        reverse(
            answer.begin(),
            answer.end()
        );

        return answer;
    }
};
*/